staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/convertlastapplied.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldmanager
+
+import (
+	"encoding/json"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal"
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/merge"
+)
+
+type convertLastAppliedManager struct {
+	fieldManager Manager
+
+	typeConverter   internal.TypeConverter
+	objectConverter runtime.ObjectConvertor
+	groupVersion    schema.GroupVersion
+}
+
+var _ Manager = &convertLastAppliedManager{}
+
+func NewConvertLastAppliedManager(fieldManager Manager, typeConverter internal.TypeConverter, objectConverter runtime.ObjectConvertor, gv schema.GroupVersion) Manager {
+	return &convertLastAppliedManager{
+		fieldManager:    fieldManager,
+		typeConverter:   typeConverter,
+		objectConverter: objectConverter,
+		groupVersion:    gv,
+	}
+}
+
+// Update implements Manager.
+func (f *convertLastAppliedManager) Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
+	return f.fieldManager.Update(liveObj, newObj, managed, manager)
+}
+
+// Apply takes field ownership of fields in the client-side apply annotation
+// so that upgrading from client-side apply to server-side apply doesn't conflict
+// by default because the user intends to manage the object with server-side apply now
+func (f *convertLastAppliedManager) Apply(liveObj runtime.Object, newObj runtime.Object, managed Managed, manager string, force bool) (runtime.Object, Managed, error) {
+	newLiveObj, newManaged, newErr := f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+
+	// Upgrade the client-side apply annotation
+	// only if the manager is the default manager for kubectl server-side apply
+	//
+	// TODO: check the user agent instead
```

Comment:
Hm I'm not sure how we should get the user-agent here since we don't have the request context. I wonder if there are alternatives

Reply:
I think I need to try passing the user agent through to the FieldManager.

It seems better than changing the `Apply` interface only for this use case.

Reply:
But the field manager is a "final" state, the userAgent is per operation

Reply:
Maybe we need to not use the layered fieldmanager for this and we do it before? I think that's fine too ...

Reply:
```golang
if userAgent == "kubectl" {
   obj = UpdateLastApplied(obj)
}
fieldManager.Apply(obj, ...)
```
```

Reply:
Ahh I see, that is viable. I'll do this

Reply:
Okay, I've tried this